proposals/0303-swiftpm-extensible-build-tools.md

```diff
@@ -293,70 +262,234 @@ protocol TargetBuildContext {
+class LibraryProduct: Product {
+    /// Whether the library is static, dynamic, or whether the type is
+    /// determined automatically.
+    let type: LibraryType
+
+    /// A type of library product.
+    enum LibraryType {
+        /// Static library.
+        case `static`
+
+        /// Dynamic library.
+        case `dynamic`
+
+        /// The type of library is unspecified and will be determined
+        /// at build time.
+        case automatic
+    }
+}
+
+/// Represents a target defined in a package. Specializations represent
+/// different types of targets.
+class Target {
+    /// The name of the target, as defined in the package manifest. It is
+    /// unique among the targets of the package.
+    let name: String
+    
+    /// The absolute path of the target directory in the local file system.
+    let directory: Path
+    
+    /// Any other targets on which this target depends, in the same order
+    /// as specified in the package manifest. Any conditional dependencies
+    /// that do not apply will have already been filtered out.
+    let dependencies: [Dependency]
+
+    /// Represents a dependency of a target on a product or on another target.
+    enum Dependency {
+        /// A dependency on a target in the same package.
+        case target(Target)
+
+        /// A dependency on a product in another package.
+        case product(Product, package: Package)
     }
 }
 
-/// Constructs commands to run during the build, including command lines,
-/// environment variables, initial working directory, etc. All paths should
-/// be based on the ones passed to the plugin in the target build context.
-protocol CommandConstructor {
+/// Represents a target consisting of a source code module, containing either
+/// Swift or source files in one of the C-based languages.
+class SourceModuleTarget: Target {
+    /// The name of the module produced by the target (derived from the target
+    /// name, though future SwiftPM versions may allow this to be customized).
+    let moduleName: String
+    
+    /// The directory containing public C headers, if applicable. This will
+    /// only be set for targets that have Clang sources, and only if there is
+    /// a public headers directory.
+    let publicHeadersDirectory: Path?
+ 
+    /// The source files that are associated with this target (any files that
+    /// have been excluded in the manifest have already been filtered out).
+    let sourceFiles: FileList
+}
+
+/// Represents a target describing a system library that is expected to be
+/// present on the host system.
+class SystemLibraryTarget: Target {
+    /// The directory containing public C headers, if applicable.
+    let publicHeadersDirectory: Path?
+}
+
+/// Represents a target describing a library that is distributed as a binary.
+class BinaryLibraryTarget: Target {
+    /// The binary library path, having a filename suffix of `.xcframeworks`.
+    let libraryPath: Path
```

Comment:
@neonichu What else should we expose to plugins about binary library targets?  Same basic question as for system library targets.

Reply:
I think the path to the library is the primary one. Not sure if it would be interesting for any plugins to know whether the binary was a remote one and its origin URL?

Reply:
Thanks!